### PR TITLE
Basic IC3 Implementation

### PR DESCRIPTION
**Pull request by makaimann**

### CODE REVIEW

**Comment by makaimann** [comment]
@ahmed-irfan Here are the latest results. It did better than I expected for not having _any_ generalization.

Screen Shot 2020-07-30 at 11 14 15 AM

File: engines/mbic3.h [hunk 1]
```diff
@@ -24,26 +24,19 @@
...
-  smt::Term term_;     // term representation of literals as disjunction
-};
-
-struct Cube
-{
-  Cube() {}
-  Cube(const smt::SmtSolver & solver, const smt::TermVec & lits);
+  Conjunction() {}
+  Conjunction(const smt::SmtSolver & solver, const smt::TermVec & lits);
   smt::TermVec lits_;  // list of literals sorted by hash
```
Comment (ahmed-irfan) [inline engines/mbic3.h:35]: this should be conjunts_

Comment (ahmed-irfan) [reply engines/mbic3.h:35]: also in the constructor

File: engines/mbic3.cpp [hunk 2]
```diff
@@ -676,7 +676,7 @@ size_t ModelBasedIC3::push_blocking_clause(size_t i, Term c)
 
   Result r;
   size_t j;
-  for (j = i; j < frames_.size(); ++j) {
+  for (j = i; j + 1 < frames_.size(); ++j) {
```
Comment (ahmed-irfan) [inline engines/mbic3.cpp:679]: can you check this @makaimann ?

File: engines/mbic3.h (hunk 1 above)
Comment (makaimann) [reply engines/mbic3.h]: Good catch! Fixed with eb808a090671e330a379b06844ee32f975a69e87

File: engines/mbic3.cpp (hunk 2 above)
Comment (makaimann) [reply engines/mbic3.cpp:679]: Yep that looks good to me! Should be equivalent to the previous fix but this is much cleaner, thanks

**Comment by makaimann** [comment]
Here are the latest runs. Still looking good. Looks like we're not adding any unsat benchmarks anymore but still picking up more sat ones.
Screen Shot 2020-08-27 at 7 49 11 AM

**Comment by makaimann** [comment]
Here are the latest results. Looks like the one-shot inductive generalization is actually worse in performance (at least on these benchmarks). The jobs are tagged with the first 5 characters of the hash, so the latest one is the `b0c40` one. I'm not sure that means it's necessarily worse for the interpolant generation though. What do you think?
Screen Shot 2020-08-28 at 8 42 47 AM

**Pull request merged**